calico_test/tests/st/utils/utils.py

```diff
@@ -33,6 +38,125 @@
+                                       (time_now % 3600) / 60,
+                                       time_now % 60)
+
+    level = kwargs.pop("level", logging.INFO)
+    msg = elapsed_hms + str(msg) % args
+    banner = "+" + ("-" * (len(msg) + 2)) + "+"
+    logger.log(level, "\n" +
+               banner + "\n"
+                        "| " + msg + " |\n" +
+               banner)
+
+
+def decorate_with_hooks(test_method):
+    """
+    Wraps each test_method with calls to the test hooks.
+    """
+
+    @functools.wraps(test_method)
+    def call_test_hooks(self):
+        """
+        Calls the test_method. If the test fails or succeeds, runs the
+        on_failure hook or the on_success hook defined in the test class.
+        """
+        try:
+            log_banner("TEST STARTING: %s", self.id())
+            test_method(self)
+        except KeyboardInterrupt:
+            log_banner("TEST INTERRUPTED: %s", self.id(),
+                       level=logging.WARNING)
+            raise
+        except SkipTest:
+            log_banner("TEST SKIPPED: %s", self.id())
+            raise
+        except Exception as e:
+            e_trace = sys.exc_info()[2]
+            logger.exception("Test %s failed with exception", self.id())
+            log_banner("TEST FAILED: %s", self.id(),
+                       level=logging.ERROR)
+            self.on_failure(test_id=self.id())
+            raise e, None, e_trace
+        else:
+            log_banner("TEST SUCCEEDED: %s", self.id())
+            self.on_success()
+        finally:
+            log_banner("TEST CLEAN-UP DONE: %s", self.id())
+
+    return call_test_hooks
+
+
+class ResultsWithHooks(type):
+    """
+    Metaclass that decorates all test_* class methods with decorate_with_hooks.
+    """
+    def __new__(cls, name, bases, dct):
+        for attr, value in dct.iteritems():
+            if attr.startswith("test_") and hasattr(value, '__module__'):
+                dct[attr] = decorate_with_hooks(value)
+        return super(ResultsWithHooks, cls).__new__(cls, name, bases, dct)
+
+
+class TestWithHooks(unittest.TestCase):
+    """
+    Base unittest class implementing success and failure hooks.
+    """
+
+    __metaclass__ = ResultsWithHooks
+
+    @classmethod
+    def on_success(cls):
+        """
+        The test completed without raising an exception.
+        """
+        logger.info("Test succeeded")
+
+    @classmethod
+    def on_failure(cls, test_id=None):
+        """
+        The test raised an exception.
+        """
+        cls.get_diags()
+        logger.info("Test failed.")
+
+    @classmethod
+    def get_diags(cls):
+        logfiles = [
+            "/var/log/calico/bird/current",
+            "/var/log/calico/bird6/current",
+            "/var/log/calico/confd/current",
+            "/var/log/calico/felix/current",
+            "/var/log/calico/libnetwork/current",
+        ]
+        if cls.hosts:
+            for host in cls.hosts:
```

Comment:
Doesn't this get invoked *after* the test has failed ... in which case the DockerHosts have exited and the containers won't be around anymore?